scipy/cluster/_hierarchy.pyx

```diff
@@ -783,6 +789,141 @@ def linkage(double[:] dists, double[:, :] Z, int n, int method):
+    method : int
+        The linkage method. 0: single 1: complete 2: average 3: centroid
+        4: median 5: ward 6: weighted
+
+    Returns
+    -------
+    Z : ndarray, shape (n - 1, 4)
+        Computed linkage matrix.
+    """
+    Z_arr = np.empty((n - 1, 4))
+    cdef double[:, :] Z = Z_arr
+
+    cdef double[:] D = dists.copy()  # Distances between clusters.
+    cdef int [:] size = np.ones(n, dtype=np.intc)  # Sizes of clusters.
+
+    cdef linkage_distance_update new_dist = linkage_methods[method]
+
+    # Variables to store neighbors chain.
+    cdef int[:] cluster_chain = np.ndarray(n, dtype=np.intc)
+    cdef int chain_length = 0
+
+    cdef int i, j, k, x, y, nx, ny, ni
+    cdef double dist, current_min
+
+    for k in range(n - 1):
+        if chain_length == 0:
+            chain_length = 1
+            for i in range(n):
+                if size[i] > 0:
+                    cluster_chain[0] = i
+                    break
+
+        # Go through chain of neighbors until two mutual neighbors are found.
+        while True:
+            x = cluster_chain[chain_length - 1]
+            current_min = NPY_INFINITYF
+
+            for i in range(n):
+                if size[i] == 0 or x == i:
+                    continue
+
+                dist = D[condensed_index(n, x, i)]
+                if dist < current_min:
+                    current_min = dist
+                    y = i
+
+            if chain_length > 1 and y == cluster_chain[chain_length - 2]:
+                break
+
+            cluster_chain[chain_length] = y
+            chain_length += 1
+
+        # Merge clusters x and y and pop them from stack.
+        chain_length -= 2
+
+        # get the original numbers of points in clusters x and y
+        nx = size[x]
+        ny = size[y]
+
+        # Record the new node.
+        Z[k, 0] = x
+        Z[k, 1] = y
+        Z[k, 2] = current_min
+        Z[k, 3] = nx + ny
+        size[x] = 0  # Cluster x will be dropped.
+        size[y] = nx + ny  # Cluster y will be replaced with the new cluster
+
+        # Update the distance matrix.
+        for i in range(n):
+            ni = size[i]
+            if ni == 0 or i == y:
+                continue
+
+            D[condensed_index(n, i, y)] = new_dist(
+                D[condensed_index(n, i, x)],
+                D[condensed_index(n, i, y)],
+                current_min, nx, ny, ni)
+
+    # Sort Z by cluster distances.
+    order = np.argsort(Z_arr[:, 2], kind='mergesort')
+    Z_arr = Z_arr[order]
+
+    # Find correct cluster labels inplace.
+    label(Z_arr, n)
+
+    return Z_arr
+
+
+cdef class LinkageUnionFind:
+    cdef int [:] parent
+    cdef int next_label
+
+    def __init__(self, int n):
+        self.parent = np.arange(2 * n - 1, dtype=np.intc)
+        self.next_label = n
+
+    cdef merge(self, int x, int y):
+        self.parent[x] = self.next_label
+        self.parent[y] = self.next_label
+        self.next_label += 1
```

Comment:
Are you referring to what is called as "union by rank" in [wikipedia](https://en.wikipedia.org/wiki/Disjoint-set_data_structure)? From the first and conceptual look: using ranks tends to create balanced trees, but in our clustering problem the tree structure is generally arbitrary, so it seems like some contradiction.

Do you see any problems with this approach? (Basically I took it from the paper and it looked very logical and straightforward for me).